src/notebooks/outputs/dataWranglerHtmlRendererComms.node.ts

```diff
@@ -0,0 +1,283 @@
+
+/**
+ * Helper to determine whether the output is a Pandas output.
+ * Specifically, we will only support the following syntaxes:
+ * 1. <df_variable_name>
+ * 2. <df_variable_name>.head(<number or variable_name>)
+ * 3. <anything>; <df_variable_name>
+ * 4. <anything>; <df_variable_name>.head(<number or variable_name>)
+ *
+ * TODO@DW: consider adding support for print(<df_variable_name>)
+ * the above uses a non-HTML format, so we would probably want to add
+ * special detection for that as well.
+ */
+export function extractVariableNameFromLine(code: string) {
+    // case 1: <df_variable_name>
+    const varMatches = code.match(new RegExp(`^(${pythonVariablePattern})\\s*$`));
+    if (varMatches?.length && varMatches.length > 1) {
+        return varMatches[1];
+    }
+    // case 2: <df_variable_name>.head(<number or variable_name>)
+    const headMatches = code.match(new RegExp(`^${dfHeadPattern}$`));
+    if (headMatches?.length && headMatches.length > 1) {
+        return headMatches[1];
+    }
+    // case 3: <anything>; <df_variable_name>
+    const multiLineVarMatches = code.match(new RegExp(`^.*;\\s*(${pythonVariablePattern})\\s*$`));
+    if (multiLineVarMatches?.length && multiLineVarMatches.length > 1) {
+        return multiLineVarMatches[1];
+    }
+    // case 4: <anything>; <df_variable_name>
+    const multiLineHeadMatches = code.match(new RegExp(`^.*;\\s*${dfHeadPattern}$`));
+    if (multiLineHeadMatches?.length && multiLineHeadMatches.length > 1) {
+        return multiLineHeadMatches[1];
+    }
+    return undefined;
+}
+
+export function getDataFrameVariableNameFromSource(source: string): string | undefined {
+    const lines = source.trim().split('\n');
+    if (lines.length === 0) {
+        return undefined;
+    }
+
+    // only look at the last line (after trimming whitespace)
+    const lastLine = lines[lines.length - 1];
+    return extractVariableNameFromLine(lastLine);
+}
+
+/**
+ * Host for the Data Wrangler entrypoint renderer.
+ */
+@injectable()
+export class DataWranglerHtmlRendererCommunicationHandler implements IExtensionSingleActivationService {
+    private readonly outputIdToVariableName: { [outputId: string]: string } = {};
+
+    constructor(
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
+        @inject(IDataScienceErrorHandler) private readonly errorHandler: IDataScienceErrorHandler,
+        @inject(IExportedKernelServiceFactory) private readonly kernelServiceFactory: IExportedKernelServiceFactory,
+        @inject(IJupyterVariables)
+        @named(Identifiers.KERNEL_VARIABLES)
+        private readonly jupyterVariables: IJupyterVariables,
+        @inject(IKernelProvider) private readonly kernelProvider: IKernelProvider
+    ) {}
+
+    public async activate(): Promise<void> {
+        const dataWranglerExtension =
+            vscode.extensions.getExtension<IDataWranglerExtensionAPI>('ms-toolsai.data-wrangler');
```

Comment:
This is a weird dependency structure.
Looks like DW depends on Jupyter and Jupyter depends on DW.
I think we can fix this if we just move DW ito a seprate extension.

Reply:
DW is a separate extension, but the renderer is here (our renderers have priority in VS code, otherwise the DW renderer is never picked).

I agree this dependency structure is weird. DW should register something with jupyter and this renderer could look for that something. Otherwise we're circular.

Reply:
It's weird that Jupyter contributes an output renderer, which is always available to users, but only functional if Data Wrangler is installed. The contribution for this renderer should only be from the Data Wrangler extension.

> our renderers have priority in VS code

It seems we don't contribute a `text/html` renderer in Jupyter? https://github.com/microsoft/vscode-notebook-renderers/blob/main/package.json

Reply:
I think this workaround is fine for now (although I agree with Don it should go in the jupyter renderers extension). We can log an issue on vscode core to allow other extensions to have priority. Not sure why they don't.

Reply:
got it, so maybe for now I can move the renderer UI into the jupyter renderers extension and keep the handling code in DW itself? e.g. expose a message listener, similar to how Jupyter is interacting with some of the renderer webviews today?

Reply:
That sounds good to me. Just to summarize:
- Renderer side code would move to jupyter renderer for now until we can resolve priority issues in VS code
- Extension side code would move into the DW extension. 
- Dependency on the DW extension would be avoided in the renderer because it would just listen for messages and just fire off another message when user clicked the button. The dependency would be the messages sent. 

@rebornix and @DonJayamanne what do you guys think?

Reply:
> Renderer side code would move to jupyter renderer for now until we can resolve priority issues in VS code

We had good discussions offline. The core would need to allow users to pick their preferred renderer for a mimetype (note that it's not the renderer extension declaring their priority).

Reply:
I think that all depends upon 'how' the user picks. There's a discoverability problem at the moment. Will there be core changes to address the discoverability problem?